doctr/models/recognition/master/pytorch.py

```diff
@@ -0,0 +1,290 @@
+        num_heads: number of heads for the mutli-head attention module
+        num_layers: number of decoder layers to stack
+        max_length: maximum length of character sequence handled by the model
+        input_size: size of the image inputs
+    """
+
+    def __init__(
+        self,
+        vocab: str,
+        d_model: int = 512,
+        headers: int = 1,
+        dff: int = 2048,
+        num_heads: int = 8,
+        num_layers: int = 3,
+        max_length: int = 50,
+        input_shape: Tuple[int, int, int] = (3, 48, 160),
+        cfg: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        super().__init__()
+
+        self.max_length = max_length
+        self.vocab = vocab
+        self.cfg = cfg
+        self.vocab_size = len(vocab)
+        self.num_heads = num_heads
+
+        self.feature_extractor = MAGCResnet(headers=headers)
+        self.seq_embedding = nn.Embedding(self.vocab_size + 1, d_model)  # One additional class for EOS
+
+        self.decoder = Decoder(
+            num_layers=num_layers,
+            d_model=d_model,
+            num_heads=num_heads,
+            dff=dff,
+            vocab_size=self.vocab_size,
+            maximum_position_encoding=max_length,
+        )
+        self.feature_pe = positional_encoding(input_shape[1] * input_shape[2], d_model)
+        self.linear = nn.Linear(d_model, self.vocab_size + 1)
+        nn.init.kaiming_uniform_(self.linear.weight)
+
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
+            elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
+                nn.init.constant_(m.weight, 1)
+                nn.init.constant_(m.bias, 0)
+
+    def make_mask(self, target: torch.Tensor) -> torch.Tensor:
+        look_ahead_mask = create_look_ahead_mask(target.size(1))
+        target_padding_mask = create_padding_mask(target, self.vocab_size)  # Pad with EOS
+        combined_mask = target_padding_mask & look_ahead_mask
+        return torch.tile(combined_mask.permute(1, 0, 2), (self.num_heads, 1, 1))
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        target: Optional[List[str]] = None,
+        return_model_output: bool = False,
+        return_preds: bool = False,
+        **kwargs: Any,
+    ) -> torch.Tensor:
+        """Call function for training
+
+        Args:
+            x: images
+            target: list of str labels
+            return_model_output: if True, return logits
+            return_preds: if True, decode logits
+
+        Return:
+            A torch tensor, containing logits
+        """
+
+        # Encode
+        feature = self.feature_extractor(x, **kwargs)
+        b, c, h, w = (feature.size(i) for i in range(4))
+        feature = torch.reshape(feature, shape=(b, c, h * w))
+        feature = feature.permute(0, 2, 1)  # shape (b, h*w, c)
+        encoded = feature + self.feature_pe[:, :h * w, :]
```

Comment:
Not sure I get it, if `feature` is of shape (N, H*W, C), how does it work that we slice the positional encoding features that way? (I mean: we may lose some information in the process if the tensor is bigger?)

Reply:
We compute the embedding matrix in the constructor (shape of the input) but then the input will be down-sized in the feature extractor and we need to keep only the n-th first embedding positions. I think that way you have more flexibility if you want to change the down sizing for instance